#1259 Add GroupPushRules and GroupPushRulesManager classes

gokeefe opened the pull request:
Add classes to read and update push rules on groups

Closes #1259.

nejch commented on gitlab/v4/objects/__init__.py:
```diff
@@ -1350,6 +1350,44 @@ class GroupProjectManager(ListMixin, RESTManager):
+    _path = "/groups/%(group_id)s/push_rule"
+    _obj_cls = GroupPushRules
+    _from_parent_attrs = {"group_id": "id"}
+    _create_attrs = (
+        tuple(),
+        (
+            "deny_delete_tag",
+            "member_check",
+            "prevent_secrets",
+            "commit_message_regex",
+            "branch_name_regex",
+            "author_email_regex",
+            "file_name_regex",
+            "max_file_size",
+        ),
+    )
+    _update_attrs = (
+        tuple(),
+        (
+            "deny_delete_tag",
+            "member_check",
+            "prevent_secrets",
+            "commit_message_regex",
+            "branch_name_regex",
+            "author_email_regex",
+            "file_name_regex",
+            "max_file_size",
+        ),
```
Could you please check the existing list of arguments available and update this list? https://docs.gitlab.com/ce/api/groups.html#add-group-push-rule just a few missing.
Thanks :) (even though one if them is missing from the response per your comment, and I realize this means our project push rules are probably outdated ... I've opened https://github.com/python-gitlab/python-gitlab/issues/1282)

gokeefe replied:
That's done. I also updated the ProjectPushRulesManager with the missing attributes. Not sure if you'd prefer that in a separate pull request.

gokeefe replied:
I'm not too familiar with pytest so i'm going to have to work on the unit tests a bit more

nejch reviewed (changes requested):
Thanks for this @gokeefe! Just added a small comment. Do you have a link for the upstream issue with the missing response?

We can't really run integration tests at the moment on EE features but if you want to give it a shot you can add some unit tests for this in `gitlab/tests/objects`. You can look at some existing examples at https://github.com/python-gitlab/python-gitlab/tree/master/gitlab/tests/objects, and use the example responses at https://docs.gitlab.com/ce/api/groups.html#push-rules for mocks.

I rebased your branch on top of master as we switched CI systems recently and this should give you better build results.

gokeefe commented:
> Thanks for this @gokeefe! Just added a small comment. Do you have a link for the upstream issue with the missing response?

Yes, right here: https://gitlab.com/gitlab-org/gitlab/-/issues/297559

> I rebased your branch on top of master as we switched CI systems recently and this should give you better build results.

Awesome, the travis-ci build failed and I had no idea why because the same tests were working fine locally. I'll have a look at the rest next week. Thanks!

nejch commented:
> > Thanks for this @gokeefe! Just added a small comment. Do you have a link for the upstream issue with the missing response?
> 
> Yes, right here: https://gitlab.com/gitlab-org/gitlab/-/issues/297559
> 
> > I rebased your branch on top of master as we switched CI systems recently and this should give you better build results.
> 
> Awesome, the travis-ci build failed and I had no idea why because the same tests were working fine locally. I'll have a look at the rest next week. Thanks!

Just FYI - we've split the objects package into modules now so this will have to be rebased and moved into `gitlab/v4/objects/push_rules.py` and the manager also imported in `groups.py`. Take a look at https://github.com/python-gitlab/python-gitlab/pull/1304 for an example on how to add it.

I'll assign this to you for now but if you need help just let me know :)

nejch commented:
Hi @gokeefe are you still interested in finishing this?

I think we're ok without pytest and I can add them later if you're not comfortable doing that. This just needs a rebase to resolve some conflicts. Let me know!

JohnVillalovos commented:
@nejch I updated this to squash it down to one commit and clean things up a little bit. If you have time to review I can work on it some more if needed.

nejch commented on docs/gl_objects/groups.rst:
```diff
@@ -376,3 +376,37 @@ Delete a group hook::
+
+    group.pushrules.create({'deny_delete_tag': True})
+
+Get group push rules (returns None is there are no push rules)::
```
(Edit: ah I realize this is also wrong in the project docs, nvm)

```suggestion
Get group push rules (returns ``None`` if there are no push rules)::
```

nejch reviewed (approved):
Thanks a lot @gokeefe for the PR and @JohnVillalovos for bringing this up to date! :)
